alg/teca_regional_moisture_flux.cxx

```diff
@@ -0,0 +1,690 @@
+#include "teca_regional_moisture_flux.h"
+
+#include "teca_cartesian_mesh.h"
+#include "teca_table.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_variant_array_impl.h"
+#include "teca_variant_array_util.h"
+#include "teca_metadata.h"
+#include "teca_array_attributes.h"
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <set>
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+#if defined(TECA_HAS_CUDA)
+#include "teca_cuda_util.h"
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/iterator/zip_iterator.h>
+
+#endif
+
+using namespace teca_variant_array_util;
+using allocator = teca_variant_array::allocator;
+
+//#define TECA_DEBUG
+
+namespace
+{
+// CPU codes
+//
+
+namespace cpu
+{
+// **************************************************************************
+template <typename coord_t>
+void transform_coordinates(coord_t *theta, coord_t *phi, coord_t *sin_phi,
+    const coord_t *lon, const coord_t *lat, unsigned long nlon, unsigned long nlat)
+{
+    // transform the cooridnates
+    coord_t rad_deg = M_PI / 180.0;
+
+    for (unsigned long i = 0; i < nlon; ++i)
+    {
+        theta[i] = lon[i] * rad_deg;
+    }
+
+    for (unsigned long j = 0; j < nlat; ++j)
+    {
+        coord_t phi_j = (90.0 - lat[j]) * rad_deg;
+
+        phi[j] = phi_j;
+
+        sin_phi[j] = sin(phi_j);
+    }
+}
+
+
+// **************************************************************************
+template <typename coord_t, typename data_t>
+void divergence(data_t *div, const data_t *ivt_u, const data_t *ivt_v,
+    const coord_t *theta, const coord_t *phi, const coord_t *sin_phi,
+    unsigned long nlon, unsigned long nlat)
+{
+    unsigned long nlatm1 = nlat - 1;
+    unsigned long nlonm1 = nlon - 1;
+
+    // compute the divergence at every point
+    for (unsigned long j = 1; j < nlatm1; ++j)
+    {
+        for (unsigned long i = 1; i < nlonm1; ++i)
+        {
+            unsigned long q = j*nlon + i;
+
+            // div = d / d theta ( ivt_theta )
+            //     + d / d phi ( ivt_phi sin(phi) )
+
+            // forward difference. using a centered difference here resulted in
+            // a larger overall error.
+            data_t dthe = theta[i + 1] - theta[i];
+            data_t dphi = phi[j + 1] - phi[j];
+
+            data_t divt_the = ivt_u[q + 1] - ivt_u[q];
+            data_t divtsin_phi =  ivt_v[q + nlon] * sin_phi[j + 1] - ivt_v[q] * sin_phi[j];
+
+            div[q] = divt_the / dthe + divtsin_phi / dphi;
```

Comment:
should there be a (1/(r_earth*sin_phi[j])) multiplication here?  If it's somehow taken care of elsewhere in the code, it might be worth adding comments here about why/how

Reply:
> should there be a (1/(r_earth*sin_phi[j])) multiplication here? If it's somehow taken care of elsewhere in the code, it might be worth adding comments here about why/how

There's also a factor of r_earth^2*sin_phi from dA that cancels with the factor you mention. After a factor of r_earth remains but I pulled this out of the integral and applied at the very end. For reference see equation 8 in the overleaf. Good idea to improve the documentation, I will add equation 8 to the doxygen.